src/seqn/trans.lisp

```diff
@@ -0,0 +1,399 @@
+
+(defmethod to-vampir ((obj seqn-nat) inputs constraints)
+  (declare (ignore constraints))
+  (list (vamp:make-constant :const (mcadr obj))))
+
+(defmethod to-vampir ((obj seqn-concat) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
+        (list (vamp:make-constant :const (+ (* (expt 2 (mcadr obj)) (vamp:const car))
+                               (vamp:const cadr))))
+        (list (vamp:make-infix :op :+
+                               :lhs (vamp:make-infix :op :*
+                                                     :lhs car
+                                                     :rhs (vamp:make-constant
+                                                           :const (expt 2 (mcadr obj))))
+                               :rhs cadr)))))
+
+(defmethod to-vampir ((obj seqn-decompose) inputs constraints)
+  (declare (ignore constraints))
+  (let* ((mcar (mcar obj))
+         (car (car inputs))
+         (rng  (vamp:make-application :func :range_n
+                                      :arguments (list (vamp:make-constant :const mcar)
+                                                       car)))
+         (lst  (list (vamp:make-constant :const (1- mcar))
+                     rng))
+         (dpth (range-depth rng)))
+    (if (typep car 'vamp:constant)
+        (if (>= (vamp:const car) (expt 2 (1- mcar)))
+            (list (vamp:make-constant :const 1)
+                  (vamp:make-constant :const (- (vamp:const car) (expt 2 (1- mcar)))))
+            (list (vamp:make-constant :const 0)
+                  car))
+        (if (= dpth 0)
+            (list (vamp:make-application :func :n_th
+                                         :arguments lst)
+                  (geb.vampir:combine
+                   (vamp:make-application :func :drop_ith
+                                          :arguments lst)))
+            (list (vamp:make-application :func :n_th
+                                         :arguments (list (vamp:make-constant :const (1- mcar))
+                                                          (geb.utils:apply-n dpth
+                                                                             (lambda (x)
+                                                                               (cadr
+                                                                                (vamp:arguments
+                                                                                 (car
+                                                                                  (vamp:arguments
+                                                                                   (cadr
+                                                                                    (vamp:arguments x)))))))
+                                                                             rng)))
```

Comment:
please fix formatting. you made the `:func` come up first, so by the time we get to `:arguments` it's already too indented for the rest of the logic. so the lambda looks like a mess, rather then a more natural formulation